### PR TITLE
Refactoring to reduce sonar code smell fixes

### PR DESCRIPTION
**Pull request by ginoaugustine**
Refactoring to reduce sonar code smell fixes

### CODE REVIEW

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java [hunk 1]
```diff
@@ -261,7 +262,7 @@ public class AnalyzerGuru {
                 new IgnorantAnalyzerFactory(),
                 new BZip2AnalyzerFactory(),
                 new XMLAnalyzerFactory(),
-                YamlAnalyzerFactory.DEFAULT_INSTANCE,
+                new YamlAnalyzerFactory(),
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:265]: why is this needed ?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/BazaarHistoryParser.java [hunk 2]
```diff
@@ -110,79 +111,81 @@ public void processStream(InputStream input) throws IOException {
...
-                        try {
-                            Date date = repository.parse(s.substring("timestamp:".length()).trim());
-                            entry.setDate(date);
-                        } catch (ParseException e) {
-                            //
-                            // Overriding processStream() thus need to comply with the
-                            // set of exceptions it can throw.
-                            //
-                            throw new IOException("Failed to parse history timestamp:" + s, e);
+            if (Objects.nonNull(entry)) {
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/BazaarHistoryParser.java:114]: what is the advantage of using `Objects.nonNull` compared to simple `!= null` check ? Also, this is adding extra level of indentation for a giant switch/case block. Can this be avoided by refactoring the block into a method or by `continue` ?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/BazaarHistoryParser.java [hunk 3]
```diff
@@ -110,79 +111,81 @@ public void processStream(InputStream input) throws IOException {
...
+                        }
+                        break;
+                    default:
+                        LOGGER.log(Level.WARNING, "Unknown parser state: {0}", state);
+                        break;
                 }
+            }
         }
 
         if (entry != null && state > 2) {
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/BazaarHistoryParser.java:191]: style-wise, I'd expect `Objects.nonNull()` to be also used here, esp. given this is the same variable.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/configuration/ConfigurationHelp.java [hunk 4]
```diff
@@ -195,13 +197,13 @@ private static Object getSampleValue(Method setter, Object defaultValue) {
     }
 
     private static Object getSampleListValue(Type genType) {
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/configuration/ConfigurationHelp.java:199]: add `Nullable` annotation for the parameter ?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java [hunk 5]
```diff
@@ -982,89 +997,84 @@ private static AnalyzerFactory findForStream(InputStream in,
...
-                fac = matcher.isMagic(content, in);
-                if (fac != null) {
-                    if (LOGGER.isLoggable(Level.FINEST)) {
-                        LOGGER.log(Level.FINEST,
-                            "''{0}'': chosen by imprecise magic: {1}",
-                            new Object[]{file, fac.getClass().getSimpleName()});
+    private static Optional<AnalyzerFactory> findFactoryUsingMatcher(byte[] content, InputStream in,
+                                                                 String file,
+                                                              boolean isPrecise) {
+        var optionalFactory = matchers.stream()
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:1014]: I wonder if this could benefit from `paralleStream()`

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java [hunk 6]
```diff
@@ -1110,7 +1120,7 @@ private static String readOpening(InputStream in, byte[] sig)
         in.mark(MARK_READ_LIMIT);
 
         String encoding = IOUtils.findBOMEncoding(sig);
-        if (encoding == null) {
+        if (Objects.isNull(encoding)) {
             // SRCROOT is read with UTF-8 as a default.
             encoding = StandardCharsets.UTF_8.name();
         } else {
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:1126]: This branch for skipping the BOM can be probably avoided by using `IOUtils.createBOMStrippedReader()` below. Just an idea (needs some thought w.r.t. buffer limit).

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java [hunk 7]
```diff
@@ -1122,53 +1132,43 @@ private static String readOpening(InputStream in, byte[] sig)
...
-        boolean postHashbang = false;
+        boolean ignoreWhiteSpace = true;
+        boolean breakOnNewLine = false;
         int r;
 
         StringBuilder opening = new StringBuilder();
         BufferedReader readr = new BufferedReader(new InputStreamReader(in, encoding), OPENING_MAX_CHARS);
+
         while ((r = readr.read()) != -1) {
-            if (++nRead > OPENING_MAX_CHARS) {
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:1133]: I'd swear that somewhere I saw a `Reader` that is limited by the amount of characters it can read. It would work well in this case, I think.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java [hunk 8]
```diff
@@ -1122,53 +1132,43 @@ private static String readOpening(InputStream in, byte[] sig)
...
             char c = (char) r;
-            boolean isWhitespace = Character.isWhitespace(c);
-            if (!sawNonWhitespace) {
-                if (isWhitespace) {
-                    continue;
-                }
-                sawNonWhitespace = true;
-            }
-            if (c == '\n') {
+            if (isNewLineOrMaxReadLimit(c, ++nRead, breakOnNewLine)) {
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:1144]: it would be good to have a unit test for this method, esp, if its signature can be simplified.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/Definitions.java [hunk 9]
```diff
@@ -183,17 +188,11 @@ public List<Tag> getTags() {
      * @return list of tags
      */
     public List<Tag> getTags(int line) {
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/Definitions.java:190]: add `@Nullable` + javadoc

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java (hunk 1 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:265]: Removed this singleton instances are harder to test, so it is recommended to avoid singleton design pattern. 
https://rules.sonarsource.com/java/RSPEC-6548/

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java (hunk 5 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:1014]: Parallel Stream have additional overhead of creating the records, splitting collection etc, so it is recommend to  user sequential streams for most of the usecases.
Parallel Stream have better performance if the collection size is huge or if we are doing many computation for each element.
Considering these factors I think sequential stream is a better fit here.
https://www.baeldung.com/java-when-to-use-parallel-stream
https://stackoverflow.com/questions/23170832/java-8s-streams-why-parallel-stream-is-slower

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/BazaarHistoryParser.java (hunk 2 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/BazaarHistoryParser.java:114]: Objects.nonNull is same as !=null, AAdvantage is while using in streams.
ex: testlList.stream().filter(Objects::nonNull) vs testlList.stream().filter(elem -> elem!=null)

Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/BazaarHistoryParser.java:114]: Skip this swtch case to multiple methods

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/BazaarHistoryParser.java (hunk 3 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/BazaarHistoryParser.java:191]: Found a way to avoid this null check

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java (hunk 6 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:1126]: createBOMStrippedReader api is not skipping BOM bytes, here we have the logic to skip the BOM bytes from the stream

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java (hunk 8 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java]: Removed this method in the latest push

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java (hunk 7 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:1133]: I was not able to find a reader with max char limit, changed the loop to for with exit condition based upon characters read.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java [hunk 10]
```diff
@@ -1109,66 +1116,68 @@ private static String readOpening(InputStream in, byte[] sig)
...
+     * @return optional of encoding
+     * @throws IOException in case of any read error
+     */
+    private static Optional<String> findAndSkipBomEncoding(InputStream in, byte[] sig) throws IOException {
+        var encoding = IOUtils.findBOMEncoding(sig);
+        if (Objects.isNull(encoding)) {
+            // SRCROOT is read with UTF-8 as a default.
+            encoding = StandardCharsets.UTF_8.name();
+        }
+        var skipForBOM = IOUtils.skipForBOM(sig);
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:1174]: `skipForBOM` will call `findBOMEncoding()` once again for the same input. Perhaps it should have a variant where `encoding` can be passed in. Not this PR, though.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/Definitions.java [hunk 11]
```diff
@@ -180,20 +186,15 @@ public List<Tag> getTags() {
...
-            result = new ArrayList<>();
-            for (Set<Tag> ltags : lineMap.sym_tags.values()) {
-                result.addAll(ltags);
-            }
-        }
-
-        return result;
+    public @Nullable List<Tag> getTags(int line) {
+        return Optional.ofNullable(lineMaps.get(line))
+                .map( lineMap -> lineMap.symTags.values().stream()
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/Definitions.java:193]:
```suggestion
                .map(lineMap -> lineMap.symTags.values().stream()
```

File: opengrok-indexer/src/main/java/org/opengrok/indexer/util/PathUtils.java [hunk 12]
```diff
@@ -135,14 +136,12 @@ public static String getRelativeToCanonical(Path path, Path canonical,
...
-                    String format = String.format("%1$s is prohibited symlink", iterPath);
-                    LOGGER.finest(format);
-                    throw new ForbiddenSymlinkException(format);
-                }
+            if (Objects.nonNull(allowedSymlinks) && Files.isSymbolicLink(iterPath) &&
+                !isWhitelisted(iterCanon.toString(), canonicalRoots) &&
+                !isAllowedSymlink(iterCanon, allowedSymlinks)) {
+                String format = String.format("%1$s is prohibited symlink", iterPath);
+                LOGGER.finest(format);
+                throw new ForbiddenSymlinkException(format);
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/util/PathUtils.java:144]:
```suggestion
                    String format = String.format("%1$s is prohibited symlink", iterPath);
                    LOGGER.finest(format);
                    throw new ForbiddenSymlinkException(format);
```

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/Repository.java [hunk 13]
```diff
@@ -86,7 +86,7 @@ public abstract class Repository extends RepositoryInfo {
      * List of &lt;revision, tags&gt; pairs for repositories which display tags
      * only for files changed by the tagged commit.
      */
-    protected TreeSet<TagEntry> tagList = null;
+    protected NavigableSet<TagEntry> tagList = null;
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/history/Repository.java:89]: what is the navigable set used for ?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/web/Laundromat.java [hunk 14]
```diff
@@ -78,25 +79,24 @@ public static String launderLog(String value) {
...
+
+        Optional.ofNullable(map)
+                .stream()
+                .map(Map::entrySet)
+                .flatMap(Collection::stream)
+                .sorted(Map.Entry.comparingByKey())
+                .forEach(entry ->
+                        safes.compute(
+                                launderLog(entry.getKey()),
+                                (k, v) -> mergeLogArrays(entry.getValue(), v)
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/web/Laundromat.java:92]: Could `v` be `null` ? Would it be better to use `safes.merge()` ?

File: opengrok-indexer/src/main/java/org/opengrok/indexer/search/SettingsHelper.java [hunk 15]
```diff
@@ -71,10 +72,9 @@ public Map<String, IndexedSymlink> getSymlinks(String projectName) throws IOExce
         getSettings(projectName);
         String projectKey = projectName != null ? projectName : "";
         Map<String, IndexedSymlink> indexSymlinks = mappedIndexedSymlinks.get(projectKey);
-        if (indexSymlinks != null) {
-            return Collections.unmodifiableMap(indexSymlinks);
-        }
-        return null;
+        return Optional.ofNullable(indexSymlinks)
+                .map(Collections::unmodifiableMap)
+                .orElseGet(Collections::emptyMap );
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/search/SettingsHelper.java:77]:
```suggestion
                .orElseGet(Collections::emptyMap);
```

**Review by vladak — CHANGES_REQUESTED** [review]
I went through all the changes, only have a bunch of comments.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java (hunk 10 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/AnalyzerGuru.java:1174]: Will take this up as a separate PR

File: opengrok-indexer/src/main/java/org/opengrok/indexer/history/Repository.java (hunk 13 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/Repository.java:89]: NavigableSet is the interface implemented by TreeSet. It is recommended to use interfaces instead of implementation.

Comment (vladak) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/history/Repository.java:89]: I see.

File: opengrok-indexer/src/main/java/org/opengrok/indexer/web/Laundromat.java (hunk 14 above)
Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/web/Laundromat.java:92]: Updated to use merge approach using Collectors.toMap

File: opengrok-indexer/src/main/java/org/opengrok/indexer/web/Laundromat.java [hunk 16]
```diff
@@ -78,47 +81,36 @@ public static String launderLog(String value) {
...
-                result[i] = launderLog(values[i]);
-            }
-        }
-        if (safeValues != null) {
-            System.arraycopy(safeValues, 0, result, i, safeValues.length);
-        }
-        return result;
     }
 
     private static String replaceAll(String value, String regex, String replacement) {
```
Comment (vladak) [inline opengrok-indexer/src/main/java/org/opengrok/indexer/web/Laundromat.java:110]: add `@Nullable` ?

Comment (ginoaugustine) [reply opengrok-indexer/src/main/java/org/opengrok/indexer/web/Laundromat.java]: Done
